posts/nyc-subway-data-with-kafka-streams-and-jhipster-part-2.md

```diff
@@ -0,0 +1,94 @@
+
+All the code of this blog post can be found on this [GitHub repository](https://github.com/Falydoor/mta-kafka) and I recommend cloning it instead of copying the code examples.
+
+# Saving measurements to InfluxDB
+
+## InfluxDB setup with the microservice
+
+Starting an InfluxDB instance can be achieve using the below Docker command:
+
+```docker run -p 8086:8086 -v $PWD:/var/lib/influxdb influxdb```
+
+A database is required to save the measurements, here is the cURL command to create a database named `mta`:
+
+```curl -G http://localhost:8086/query --data-urlencode "q=CREATE DATABASE mta"```
+
+The microservice can now connect to the instance and starts saving measurements in the `mta` database using the Java driver. In order to use the Java driver, the below Maven dependency must be added first:
+
+```xml
+<dependency>
+    <groupId>org.influxdb</groupId>
+    <artifactId>influxdb-java</artifactId>
+    <version>2.14</version>
+</dependency>
+```
+
+Configuring the java client is very simple and the interface `InfluxDB` will be later used to query the database.
+
+```java
+// Init influxDB
+InfluxDB influxDB = InfluxDBFactory.connect("http://127.0.0.1:8086", "root", "root");
+influxDB.setDatabase("mta");
+```
+
+## Measurements saving
+
+In the [part 1](https://blog.ippon.tech/nyc-subway-data-with-kafka-streams-and-jhipster-part-1/), the Kafka topic `mta-stream` was used to store the results from the streaming process. This topic will be used to save each messages in InfluxDB since a message represents the number of active subways for a given line and time.
+
+```java
+@StreamListener(MessagingConfiguration.MtaStream.INPUT)
+public void saveSubwayCount(SubwayCount subwayCount) {
+    // Save measurement in influxdb
+    influxDB.write(Point.measurement("line")
+        .time(subwayCount.getStart().toEpochMilli(), TimeUnit.MILLISECONDS)
+        .tag("route", subwayCount.getRoute())
+        .addField("count", subwayCount.getCount())
+        .build());
+}
+```
+
+The method above get triggered when a message is published in the topic `mta-stream` and will then write the message in InfluxDB. Spring Cloud Stream makes things simple by converting the Kafka message from a JSON format to the custom `SubwayCount` class.
+
+InfluxDB's measurement is conceptually similar to a table and all messages are inserted in the `line` measurement. The subway's route is used as a tag to allow filtering when it will be used in Grafana. There is only one field for the point and it is the number of active subways. Finally the time of the point is the start timestamp of the Kafka's window.
```

Comment:
```suggestion
InfluxDB's measurement is conceptually similar to a table and all messages are inserted in the `line` measurement. The subway's route is used as a tag to allow filtering when it will be used in Grafana. There is only one field for the point and it is the number of active subways. Finally the time of the point is the start timestamp of Kafka's window.
```